lib/livebook/runtime/evaluator.ex

```diff
@@ -918,15 +933,122 @@ defmodule Livebook.Runtime.Evaluator do
     Enum.reject(code_markers, &(&1.line == 0))
   end
 
-  defp extra_diagnostic?(%SyntaxError{}), do: true
-  defp extra_diagnostic?(%TokenMissingError{}), do: true
-  defp extra_diagnostic?(%MismatchedDelimiterError{}), do: true
+  @compile {:no_warn_undefined, {Pythonx, :eval, 2}}
+  @compile {:no_warn_undefined, {Pythonx, :decode, 1}}
 
-  defp extra_diagnostic?(%CompileError{description: description}) do
-    not String.contains?(description, "(errors have been logged)")
+  defp eval_python(code, binding, env) do
+    with :ok <- ensure_pythonx() do
+      {result, _diagnostics} =
+        Code.with_diagnostics([log: true], fn ->
+          try do
+            quoted = python_code_to_quoted(code)
+
+            {value, binding, env} =
+              Code.eval_quoted_with_env(quoted, binding, env, prune_binding: true)
+
+            result = {:ok, value, binding, env}
+            code_markers = []
+            {result, code_markers}
+          catch
+            kind, error ->
+              code_markers =
+                if is_struct(error, Pythonx.Error) do
+                  Pythonx.eval(
+                    """
+                    import traceback
+
+                    if traceback_ is None:
+                      diagnostic = None
+                    elif isinstance(value, SyntaxError):
+                      diagnostic = (value.lineno, "SyntaxError: invalid syntax")
+                    else:
+                      description = " ".join(traceback.format_exception_only(type, value)).strip()
+                      diagnostic = (traceback_.tb_lineno, description)
+
+                    diagnostic
+                    """,
+                    %{
+                      "type" => error.type,
+                      "value" => error.value,
+                      "traceback_" => error.traceback
+                    }
+                  )
+                  |> elem(0)
+                  |> Pythonx.decode()
+                  |> case do
+                    nil -> []
+                    {line, message} -> [%{line: line, description: message, severity: :error}]
+                  end
+                else
+                  []
+                end
+
+              result = {:error, kind, error, []}
+              {result, code_markers}
+          end
+        end)
+
+      result
+    end
   end
 
-  defp extra_diagnostic?(_error), do: false
+  defp python_code_to_quoted(code) do
+    # We expand the sigil upfront, so it is not traced as import usage
+    # during evaluation.
+
+    quoted = {:sigil_PY, [], [{:<<>>, [], [code]}, []]}
+
+    env = Code.env_for_eval([])
+
+    env =
+      env
+      |> Map.replace!(:requires, [Pythonx])
+      |> Map.replace!(:macros, [{Pythonx, [{:sigil_PY, 2}]}])
+
+    Macro.expand_once(quoted, env)
+  end
+
+  defp eval_pyproject_toml(code, binding, env) do
+    with :ok <- ensure_pythonx() do
+      quoted = {{:., [], [{:__aliases__, [alias: false], [:Pythonx]}, :uv_init]}, [], [code]}
```

Comment:
Where did we land on the discussion between using the application environment for the install vs an explicit call to `uv_init`? If we could simply add it as a dep, it would be better, but I don't recall the details. :)

Reply:
Oh, I think we had to do it after the fact, hence why we pretend this to be its own cell. :+1: